### PR TITLE
Protein dl

### PR DESCRIPTION
**Pull request by KalinNonchev**

### CODE REVIEW

File: tests/test_5_protein_dl.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+def gtf_row2interval(row):
+    """Note: GTF is 1-based
+    """
+    return Interval(str(row.Chromosome),
+                    int(row.Start) - 1,
+                    int(row.End),
+                    strand=str(row.Strand))
+
+
+class GenomeCDSSeq:
```
Comment (MuhammedHasan) [inline tests/test_5_protein_dl.py]: name to GenomeCDSFetcher

File: tests/test_5_protein_dl.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+        anchor = 1
+        for transcript_id in tqdm(self.gps.transcripts):
+            yield self.extract(transcript_id, anchor)
+
+
+class SingleVariantAminoAcidVCFSeqExtractor(AminoAcidVCFSeqExtractor):
+
+    # function for a concrete protein id
+
+    def extract(self, transcript_id, anchor, sample_id=None):
```
Comment (MuhammedHasan) [inline tests/test_5_protein_dl.py]: change your algorithm and yield list

File: tests/test_5_protein_dl.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+class SingleSeqAminoAcidVCFSeqExtractor(AminoAcidVCFSeqExtractor):
+
+    # function for a concrete protein id
+
+    def extract(self, transcript_id, anchor, sample_id=None):
+        intervals, strand = self.gps.get_cds_exons(transcript_id)
+        seq = ""
+        for interval in intervals:
+            interval = Interval(interval[0], int(interval[1]), int(interval[2]))
+            seq = "".join([seq, self.sse.extract(interval, anchor=anchor)])
```
Comment (MuhammedHasan) [inline tests/test_5_protein_dl.py]: seq = ''.join([
  self.sse.extract(interval, anchor=anchor)]
   for interval in intervals
])

File: tests/test_5_protein_dl.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        while(len(seq) % 3 != 0):
+            seq = seq[:-1]
+        return seq
+
+
+class SingleSeqAminoAcidVCFSeqExtractor(AminoAcidVCFSeqExtractor):
+
+    # function for a concrete protein id
+
+    def extract(self, transcript_id, anchor, sample_id=None):
```
Comment (MuhammedHasan) [inline tests/test_5_protein_dl.py]: use `sample_id`

File: tests/test_5_protein_dl.py [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+        intervals, strand = self.gps.get_cds_exons(transcript_id)
+        seq = ""
+        for interval in intervals:
+            interval = Interval(interval[0], int(interval[1]), int(interval[2]))
```
[the remark this answers is not in the document]
Comment (KalinNonchev) [reply tests/test_5_protein_dl.py]: yes, they are different types of interval

File: tests/test_5_protein_dl.py [hunk 6]
```diff
@@ -0,0 +1,351 @@
+"""Test protein dataloader
+"""
+from pybedtools import Interval
```
Comment (MuhammedHasan) [inline tests/test_5_protein_dl.py]: change `Interval` class to kipoiseq Interval by `from kipoiseq import Interval`

File: tests/test_5_protein_dl.py [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+            seq = rc_dna(seq)
+        # optionally reverse complement
+        seq = self.genome_cds_fetcher.cut_seq(seq)
+        
+        return translate(seq)
+
+    def coding_single_seq(self):
+        for transcript_id in tqdm(self.genome_cds_fetcher.transcripts):
+            yield self.extract(transcript_id)
+
```
Comment (MuhammedHasan) [inline tests/test_5_protein_dl.py]: def extract_multiple(self, variant_interval_queryable: VariantIntervalQueryable, sample_id):
   return seq

https://github.com/kipoi/kipoiseq/blob/83f63a549716b3d031b28b9451b5870f223be17a/kipoiseq/extractors/vcf.py#L56

check size of the variants warn user if indel

and use `VariantSeqExtractor` to inject variants https://github.com/kipoi/kipoiseq/blob/83f63a549716b3d031b28b9451b5870f223be17a/kipoiseq/extractors/vcf_seq.py#L50

**Comment by brechtmann** [comment]
@Avsecz @MuhammedHasan should we copy the 'read_fasta' function from concise to kipoiseq or add concise as a dependency? Or use another implementation?

https://github.com/KalinNonchev/kipoiseq/blob/1ec039a9af0ef9d55f0191033afbfc772e781f79/tests/test_5_protein_dl.py#L5

**Comment by brechtmann** [comment]
And where should we move the classes from the test file.
should we make a file 'kipoiseq/dataloaders/protein.py' any suggestions?

**Comment by MuhammedHasan** [comment]
I will handle remaining issues.

**Comment by brechtmann** [comment]
You have two more cases of format strings which will need to be changed and discuss with Hasan how to handle the data used for testing from ensemble.
Maybe we can keep a script which can run the entire test somewhere but not run it at every CI step?

**Pull request closed without merging**